File: nextctl-go/cmd/nextctl/nextctl.go
/*
* SPDX-License-Identifier: BSD 2-Clause "Simplified" License
*
* nextctl.go
*
* Created by:	Aakash Sen Sharma, December 2022
* Copyright:	(C) 2022, Aakash Sen Sharma & Contributors
 */

package main

import (
	"fmt"
	"os"

	. "git.sr.ht/~shinyzenith/nextctl/pkg/next_control"
	"github.com/rajveermalviya/go-wayland/wayland/client"
)

const EXIT_FAILURE = 1
const EXIT_SUCCESS = 0
const VERSION = "0.1.0-dev"

type Nextctl struct {
	display      *client.Display
	registry     *client.Registry
	next_control *NextControlV1
}

const usage = `Usage: nextctl <command>
  -h, --help      Print this help message and exit.

  -v, --version   Print the version number and exit.

Complete documentation for recognized commands can be found in
the nextctl(1) man page.`

func main() {
	for _, arg := range os.Args[1:] {
		if arg == "-h" || arg == "--help" {
			fmt.Fprintln(os.Stderr, usage)
			os.Exit(0)
		} else if arg == "-v" || arg == "--version" {
			fmt.Fprintln(os.Stderr, "Nextctl version: ", VERSION)
			os.Exit(0)
		}
	}

	Nextctl := &Nextctl{display: nil, registry: nil, next_control: nil}

	if display, err := client.Connect(""); err != nil {
		fmt.Fprintln(os.Stderr, "ERROR: Cannot connect to wayland display.")
		os.Exit(EXIT_FAILURE)
	} else {
		Nextctl.display = display
	}

	Nextctl.registry, _ = Nextctl.display.GetRegistry()
	Nextctl.registry.AddGlobalHandler(Nextctl.GlobalHandler)

	Nextctl.DisplayDispatch()

	if Nextctl.next_control == nil {
		fmt.Fprintln(os.Stderr, "ERROR: Compositor doesn't implement NextControlV1.")
		os.Exit(EXIT_FAILURE)
	}

	for _, arg := range os.Args[1:] {
		_ = Nextctl.next_control.AddArgument(arg)
	}
	callback, _ := Nextctl.next_control.RunCommand()
	callback.AddSuccessHandler(Nextctl.NextSuccessHandler)
	callback.AddFailureHandler(Nextctl.NextFailureHandler)
	Nextctl.DisplayDispatch()

	os.Exit(EXIT_SUCCESS)
}

func (Nextctl *Nextctl) DisplayDispatch() {
	callback, err := Nextctl.display.Sync()
	if err != nil {
		fmt.Fprintln(os.Stderr, "ERROR: wayland dispatch failed.")
		os.Exit(EXIT_FAILURE)
	}
	defer func() {
		if err := callback.Destroy(); err != nil {
			fmt.Fprintln(os.Stderr, "ERROR: wayland dispatch failed.")
			os.Exit(EXIT_FAILURE)
		}
	}()

	callback_done := false
	callback.AddDoneHandler(func(_ client.CallbackDoneEvent) {
		callback_done = true
	})

	for !callback_done {
		Nextctl.display.Context().Dispatch()
	}
}

func (Nextctl *Nextctl) GlobalHandler(wl_global client.RegistryGlobalEvent) {
	if wl_global.Interface == "next_control_v1" {
		next_control := NewNextControlV1(Nextctl.display.Context())
		err := Nextctl.registry.Bind(wl_global.Name, wl_global.Interface, wl_global.Version, next_control)
		if err != nil {
			fmt.Fprintln(os.Stderr, "ERROR: wayland dispatch failed.")
			os.Exit(EXIT_FAILURE)
		}
		Nextctl.next_control = next_control
	}
}

func (Nextctl *Nextctl) NextSuccessHandler(success_event NextCommandCallbackV1SuccessEvent) {
	fmt.Fprintln(os.Stdout, success_event.Output)
}

func (Nextctl *Nextctl) NextFailureHandler(failure_event NextCommandCallbackV1FailureEvent) {
	fmt.Fprint(os.Stderr, "ERROR: ", failure_event.FailureMessage)
	if failure_event.FailureMessage == "Unknown command\n" || failure_event.FailureMessage == "No command provided\n" {
		fmt.Fprintln(os.Stderr, usage)
	}
}

File: nextctl/src/nextctl.c
/* SPDX-License-Identifier: BSD 2-Clause "Simplified" License
 *
 * nextctl/nextctl.c
 *
 * Created by:	Aakash Sen Sharma, May 2022
 * Copyright:	(C) 2022, Aakash Sen Sharma & Contributors
 */

#include "nextctl.h"

int main(int argc, char *argv[]) {
	for (int i = 0; i < argc; i++) {
		if (strcmp("-h", argv[i]) == 0 || strcmp("--help", argv[i]) == 0) {
			(void)fputs(usage, stderr);
			return EXIT_SUCCESS;
		} else if (strcmp("-v", argv[i]) == 0 || strcmp("--version", argv[i]) == 0) {
			(void)printf("Nextctl version: %s\n", VERSION);
			return EXIT_SUCCESS;
		}
	}

	struct nextctl_state state = { 0 };
	state.wl_display = wl_display_connect(NULL);

	if (state.wl_display == NULL) {
		(void)fputs("ERROR: Cannot connect to wayland display.\n", stderr);
		return EXIT_FAILURE;
	}

	state.wl_registry = wl_display_get_registry(state.wl_display);
	(void)wl_registry_add_listener(state.wl_registry, &registry_listener, &state);

	if (wl_display_dispatch(state.wl_display) < 0) {
		(void)fputs("ERROR: wayland dispatch failed.\n", stderr);
		return EXIT_FAILURE;
	}

	if (state.next_control == NULL) {
		(void)fputs("ERROR: Compositor doesn't implement NextControlV1.\n", stderr);
		return EXIT_FAILURE;
	}

	for (int i = 1; i < argc; i++) {
		(void)next_control_v1_add_argument(state.next_control, argv[i]);
	}

	state.next_command_callback = next_control_v1_run_command(state.next_control);
	(void)next_command_callback_v1_add_listener(state.next_command_callback,
												&callback_listener, NULL);

	if (wl_display_dispatch(state.wl_display) < 0) {
		(void)fputs("ERROR: wayland dispatch failed.\n", stderr);
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}

static void registry_handle_global(void *data, struct wl_registry *registry,
								   uint32_t name, const char *interface,
								   uint32_t version) {
	struct nextctl_state *state = data;

	if (strcmp(interface, next_control_v1_interface.name) == 0) {
		state->next_control =
			wl_registry_bind(registry, name, &next_control_v1_interface, 1);
	}
}

static void next_handle_success(void *data, struct next_command_callback_v1 *callback,
								const char *output) {
	(void)fputs(output, stdout);
}

static void next_handle_failure(void *data, struct next_command_callback_v1 *callback,
								const char *failure_message) {
	(void)fprintf(stderr, "ERROR: %s", failure_message);
	if (strcmp("Unknown command\n\0", failure_message) == 0 ||
		strcmp("No command provided\n\0", failure_message) == 0) {
		(void)fputs(usage, stderr);
	}
}
